harness/tests/tensorboard/test_util.py

```diff
@@ -88,6 +89,8 @@ def test_list_tb_files_nonexistent_directory(tmp_path: pathlib.Path) -> None:
     assert not pathlib.Path(base_path).exists()
     assert manager.list_tb_files(0, lambda _: True) == []
 
+    manager.close()
```

Comment:
Probably we need `with SharedFSTensorboardManager(...) as manager:`, or at least a try/finally for `manager.close()`.  Otherwise when these tests fail, it will cause the python interpreter to hang on shutdown waiting for the threads to get cleaned up.

Reply:
I changed the default async_upload to False and only enable it when it is used inside of core context, which is the use case we are solving for. Therefore, other existing use cases no longer have behavior change (still synchronous) and this change is no longer needed.

In the future, we can enable async_upload on a case by case basis if needed.

Reply:
I think literally all training goes through the core api (all Trial classes use core api), so I don't know if that's a great default in this case.

Reply:
The thing I like the most is making all the calls to user-provided functions in a single-threaded context; I think having the async code be as dumb as possible improves the understandability of the system from the user's perspective.